common/djangoapps/course_modes/admin.py

```diff
@@ -51,10 +53,23 @@ class Meta(object):
     )
 
     def __init__(self, *args, **kwargs):
+        # If args is a QueryDict, then the ModelForm addition request came in as a POST with a course ID string.
+        # Change the course ID string to a CourseLocator object by copying the QueryDict to make it mutable.
+        if len(args) > 0 and 'course' in args[0] and isinstance(args[0], QueryDict):
+            args_copy = args[0].copy()
+            args_copy['course'] = CourseKey.from_string(args_copy['course'])
+            args = [args_copy]
+
         super(CourseModeForm, self).__init__(*args, **kwargs)
 
-        if self.data.get('course'):
-            self.data['course'] = CourseKey.from_string(self.data['course'])
+        try:
+            if self.data.get('course'):
+                self.data['course'] = CourseKey.from_string(self.data['course'])
+        except AttributeError:
+            # Change the course ID string to a CourseLocator.
+            # On a POST request, self.data is a QueryDict and is immutable - so this code will fail.
+            # However, the args copy above before the super() call handles this case.
+            pass
 
```

Comment:
JFTR, I'm not thrilled with the code above (I wrote it).

Reply:
Would it be better to just always pre-emptively pass a mutable copy of the data parameter to the superclass `__init__`?

Reply:
I'm unclear on how that's different than what we do currently.  It seems to me that we create a mutable copy before passing to the superclass already.

Reply:
Only in one particular case.  What's here should work, I was just responding to John's concern with a possible simplification where we just always copy the input data and pre-process the "course" entry if present, clearing out a fair amount of conditional code (at the expense of a slight performance hit).